### DISCUSSION TITLE
Experience porting to other Lisps

### DISCUSSION THREAD

**Discussion by stylewarning** [coalton-lang/coalton#160] [Other]
So far I've tried to load Coalton on ECL and CLISP, and both fail. (Both stock on Debian Buster.)

ECL fails to load Serapeum, complaining about atomics not being supported or something.

CLISP fails to load Trivia because it claims there's a misplaced declaration.

**Comment by AthenaEryma** [comment]
Is Serapeum even really necessary? A quick Github search indicates that the only symbols used out of it are `defstruct-read-only` and `true`. While `true` is pretty trivial, `defstruct-read-only` is not... but Serapeum is a pretty heavyweight library, pulling in quite a few transitive dependencies, on top of the portability issues you've found. It might be worth pulling `defstruct-read-only` out to ditch the Serapeum dependency.

**Reply by stylewarning** [reply]
It's probably not necessarily. It was useful during development. We have run into Serapeum bugs, but they were hastily fixed.

We'd welcome a PR to remove it!

**Reply by eliaslfox** [reply]
Serapeum was remove in #259.

**Comment by toonnyy8** [comment]
When I try to load Coalton on ECL it still fails, and when compiled to coalton-impl/parser/read it shows:

```
;;; Error:
;;; in file toplevel.lisp, position 20672
;;; at (DEFUN READ-PROGRAM ...)
;;; * In (LAMBDA READ-PROGRAM...), the type of the form (ERROR "you must supply this bozo!") is NULL, not (MEMBER :FILE :TOPLEVEL-MACRO :TEST).
Condition of type: COMPILE-FILE-ERROR
COMPILE-FILE-ERROR while compiling #<cl-source-file "coalton/compiler" "parser" "toplevel">
```

This error can be avoided by commenting out line 436 of toplevel.lisp:
```lisp
(defun read-program (stream file &key (mode (error "you must supply this bozo!")))
  "Read a PROGRAM from the COALTON-FILE."
  (declare (type coalton-file file)
           ;; (type (member :file :toplevel-macro :test) mode)
           (values program))
  ...
```

However, when it continues to compile to coalton-library/types, a new error is displayed, but my shallow CL knowledge is not enough to solve it:
```
;;; Internal error:
;;;   ** The readmacro #<compiled-function COALTON-IMPL/READER::READ-COALTON-TOPLEVEL-OPEN-PAREN 0x7f8ac9af2480> returned 2 values.
Condition of type: COMPILE-FILE-ERROR
COMPILE-FILE-ERROR while compiling #<cl-source-file "coalton/library" "types">
```
